os/fs/smartfs/smartfs_utils.c

```diff
@@ -1833,7 +1833,102 @@ int smartfs_truncatefile(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *e
 
 int smartfs_shrinkfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
 {
-	/* TODO This will be added later */
+	int ret;
```

Comment:
Below logic is what I thought before, please see this.
If CONFIG_SMARTFS_USE_SECTOR_BUFFER enabled, sf->buffer consist of data from target sector... it is does not set, smartfs read only header, so it may more faster.
```
smartfs_seek_internal(fs, sf, length, SEEK_SET);
#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
    memset(&sf->buffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
    .....
    sf->bflags = SMARTFS_BFLAG_DIRTY;
    sync_internal(fs, sf);
#else
    smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (FAR uint8_t *)fs->fs_rwbuffer);
    ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
    memset();
    ....
    ret =  FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
    nextsector = next_sector_of_sf->currsector;
}
while(nextsector != SMARTFS_ERASEDSTATE_16BIT) {
    FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)nextsector);
    nextsector = next_sector_of_cursector;
};

Reply:
Using pre-existing seek_internal function allows reading only headers instead of entire sectors, hence preferable.
Syncing shrunk file state to MTD for better robustness against power downs.
However, need to move sync_internal and seek_internal to utils.c with definitions added to smatfs.h

Reply:
Filepos after completion of file shrink needs to be determined,
Modifications pending